### ISSUE TITLE
Backup Feature

### ISSUE THREAD

**Issue by eals2019** [firefly-iii/firefly-iii#7081]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/firefly-iii/about-firefly-iii/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been discussed before.

### Description

First, I want to apologize. I found a ticket related to this, but I still want to suggest this as a feature and explain why. I am super thrilled to use FireFly-III. For nearly two years, I've been developing a spreadsheet that works similar to FireFly, but far less feature-riched. As I've learned about databases, I've realized I'm essentially creating one in a spreadsheet. The problem, though, is that I have to essentially redesign the entire spreadsheet to add assets, piggy-banks, etc.

I decided to try out FireFly in a Ubuntu Server VM to see how it works and if I would like it, particularly to see if it was the database version of my spreadsheet. I put all my transactions for the last two months into it, and it's amazing! So much so that I would like to migrate it to a physical machine to use permanently. **However, there doesn't appear to be any good way to export the data from the program and import it into another instance of it.** phpMyAdmin has a similar feature to export a database in SQL, and it's super easy to migrate the data to a different MySQL database.

### Solution

**What I would love to see is a way of downloading all the transactions, piggy-bank information, etc. (essentially all the data FireFly works with) so that I could upload it to a new instance of the application**. As it is, my only options are to backup using Terminal, or re-enter all my data into a new instance of the program. I think a better way of exporting all the data in case the server catastrophically failed so that it could be restored onto a new installation of the program would be a super useful feature.

Based on related [tickets](https://github.com/firefly-iii/firefly-iii/issues/4270), _I realize you aren't a huge fan of this, but let me explain why I want to push for it._ FireFly-III requires a lot of technical knowledge to get setup (...I'm thinking about people who don't even know what Linux is), but once it is setup it is somewhat easy to use once you get over the learning curve of how it works. I can imagine myself setting FireFly up for my wife, parents, friends, etc., people who would benefit from this, but don't have the technical knowledge to do this themselves. **Having a more user friendly means of backing up FireFly data would allow less-technologically savvy people to backup their own, rather than relying on a tech expert to do it manually or set it up automatically**. Then, the tech expert setting up the server and application can rely on the users to backup their data.

### What are alternatives?

I don't really see any great alternatives. To me, Terminal command lines or automated backups that way doesn't seem like the most user friendly way to backup data. _It is a solution for users with vast background knowledge, but what about those of us without that knowledge_. It requires the user to have enough knowledge in Linux and Docker to perform this, which means users who had a tech expert set this up would would have to trust that the setup expert had considered backups.

As for me, I can probably (hopefully) figure out backing it up, but if I hadn't set it up for myself or couldn't figure it out, I would want to have the ability to back it up on my own through the application.

**Comment by JC5** [comment]
Hey, thanks for an excellent ticket. Let me rip off the band-aid right now, Firefly III will not get a native backup/restore feature. 

You have to realize that I'm not really interested in making Firefly III user friendly, easy to install or accessible to people who don't have the technical knowhow to install it otherwise. Up to a point I want it to work, and work correctly of course. But this is just one of those limits. 

Writing a backup and restore routine isn't interesting for me to make. Databases can do this already, why invent the wheel again? And sure somebody else could build it. Leaving me with the maintenance and support burden, no thank you. 

Use Docker and its extensive ecosystem of tools (including those for backup and restoration) makes it easy enough to do this even when you don't know anything about SQL. 

The reason I wont do this is fairly straightforward. At the end of the day, Firefly III has a target audience of one single user: me. I started building Firefly III for a single user (me) and I'm happy to continue even if Firefly III's user base drops back down to a single user again. And that single user is just not interested in having a native backup/restore solution because he knows how to backup databases. 

I'm sorry I'm not telling you what you'd like to hear but I hope it helps nonetheless.

**Comment by eals2019** [comment]
Hey,

That's completely understandable. When I first started creating my
spreadsheet, that's how it was. I created documentation for my college
class in case anyone wanted to use it, but at the end of the day it was
just for me. I get that.

In case no one told you, you have an incredible application. Last semester,
I took a database course and I'm currently in the process of developing one
for a customer as a capstone project for college, and I started to realize
my spreadsheet was in essence a database. I realized that to get the
functionality I needed, I would have to change my spreadsheet into a
database, but I simply don't have the coding skills to develop the website
and server application. FireFly turned out to be exactly what I was looking
for. Thank you for creating this and sharing it with me. Don't sell
yourself short though!

I'll have to learn more about Docker as I know nothing about it. I
anticipate I'll figure out how to restore it on a new system.
